### ISSUE TITLE
Support selecting all text on triple click

### ISSUE THREAD

**Issue by provokateurin** [flutter/flutter#63576]
## Use case
On native platforms and browsers using a triple click (with a mouse) on text in a textfield one selects all text in the textfield. Flutter doesn't do this, but rather deselects the selected word.

## Proposal
Flutter should support this to match native platforms and browsers.

**Comment by HansMuller** [comment]
CC @justinmc

**Comment by justinmc** [comment]
Thanks for pointing this out!  I agree Flutter should support this.

Two details I want to call attention to:

  1. I don't see native Android apps producing this behavior. I tried on a physical Pixel 2 running Android 10. Does anyone see this on Android?
  1. On native iOS and web, it appears to select the entire paragraph, but not all text.

**Comment by dave** [comment]
I noticed this within a few minutes of playing with a flutter app on my iPhone, so I'd say it's a significant issue. 

Can't speak for Android but as you say the usual behaviour on iOS is to select the paragraph.

**Comment by dave** [comment]
I also note that a double tap followed by a drag doesn't replicate the native iOS behaviour. In native apps, this gesture selects a block of text.

**Comment by lukepighetti** [comment]
Triple click doesn't work as expected on flutter web macOS

**Comment by mdebbar** [comment]
This is something that needs to be done on the framework side for iOS and web.

**Comment by Jon-Salmon** [comment]
As an extension to this issue, the 'triple click and drag' action should be implemented. On Window expected behaviour is that it should set the drag start (or end) point to the start (or end) of the word clicked on.

On Flutter desktop, the drag start/end point is set where the point of the click rather than the word boundary.

**Comment by venkatd** [comment]
Sharing the feedback of a beta user of our app that stopped using the beta app after not having expected behavior with text selection:

![image](https://user-images.githubusercontent.com/220686/112688665-2e90dd00-8e47-11eb-80c8-da557801f6d2.png)

@mdebbar could you share why web implementation involves the framework side? This this logic implemented on that side for performance/latency reasons?

**Comment by lukepighetti** [comment]
I feel like we need an entire initiative around improving text field performance in Flutter. We still have issues on iOS text fields from 2-3 years ago that haven't been addressed.

**Comment by justinmc** [comment]
I'm not sure why this needs to be done in the framework for web, but I think it easily could be.  This is how I think we can handle this:

1. Add support for triple taps in TextSelectionGestureDetector with a timer ([near here](https://github.com/flutter/flutter/blob/a603714610f8c9c831e7d85071274460d70c18da/packages/flutter/lib/src/widgets/text_selection.dart#L1393)).
2. When a triple tap is detected, invoke a new Intent with [Actions.invoke](https://master-api.flutter.dev/flutter/widgets/Actions/invoke.html).
3. Receive the Intent in DefaultTextEditingActions.
4. Create a method in RenderEditable to select the paragraph at a given location, and call this method from DefaultTextEditingActions.

This should work for whatever platforms we want or don't want, including web, and it would be overrideable by the app developer.

**Comment by venkatd** [comment]
@justinmc I like the idea of keeping this platform agnostic.

Do you have a sense of what the priority of this issue is? If it's not going to get any focus for 1+ months, I'm happy to give this a try myself or at least get it started.

I may need some direction though.

Some questions:
* How do we disambiguate a triple click from a double click? Is there some kind of debounce delay we have before registering single vs. double vs. triple click? 
* What defines a paragraph? Is there existing logic in Flutter to return a paragraph given a cursor position and text?

**Comment by justinmc** [comment]
I think you should go for it.  I'm trying to convert all interactions like this to use Actions, but I'm focused on the existing ones for now.

1. Take a look at [_isWithinDoubleTapTolerance](https://github.com/flutter/flutter/blob/a603714610f8c9c831e7d85071274460d70c18da/packages/flutter/lib/src/widgets/text_selection.dart#L1495).  I think we'd need to implement a similar thing for triple click.  If I'm not mistaken, Flutter has nothing built-in for detecting triple clicks/taps.
2. I couldn't find anything built-in for this with a quick glance.  Looking for something kind of like [_getLineAtOffset](https://github.com/flutter/flutter/blob/f12d90c4358be1d0ec159c9f72ea3ccc02fd9f59/packages/flutter/lib/src/rendering/editable.dart#L3077) in RenderEditable or TextPainter.  However, I think really what we want is just the text between the two nearest newlines, which should be simple enough to find, even if it's not really performant for giant paragraphs.

**Comment by lukepighetti** [comment]
Seems like Flutter should just have a AnyNumberClick detector. I have built this for other projects and it wasn't terribly complex.

**Comment by venkatd** [comment]
Great, I'll give it a go.

I'll go for something basic with a `_getParagraphAtOffset` implementation and we can improve it after.

@lukepighetti good point. `onDoubleTap` should fire the proper select word behavior and `onTripleTap` should extend the selection to a paragraph. I noticed that if I double tap, pause a second, then tap one more, it deselects. So the triple tap has some sort of expiration.

I'll try to get a starting point working, then iterate on the details. Hopefully will have time later in the week to take this on.

**Comment by justinmc** [comment]
Sounds good!  I think that in the scope of this PR we shouldn't worry about creating a generic onTripleTap in GestureDetector (though it might be a good idea for another PR).  Currently for text selection, double tap is implemented from scratch and doesn't use GestureDetector's built-in onDoubleTap (see [here](https://github.com/flutter/flutter/blob/a603714610f8c9c831e7d85071274460d70c18da/packages/flutter/lib/src/widgets/text_selection.dart#L1393)) because it needs to respond to both the single and double tap down events.  That's actually probably good news for us because it should be easier to extend to handle a third click.

**Comment by lukepighetti** [comment]
One thing we might not be able to do is be aware of the user's system settings for double click times. Maybe that's why it was suggested that this is above the Flutter widget layer?

**Comment by justinmc** [comment]
Ah I didn't even think of system settings for double tap time.  It looks like it's hardcoded in Flutter now ([code](https://github.com/flutter/flutter/blob/89f82ff4fb46b5bcbf57d3fdd6e34d2d72993fd8/packages/flutter/lib/src/gestures/multitap.dart#L237)).  Long term we should investigate getting that information from the engine, but for now I think we can just keep it hardcoded.

**Comment by justinmc** [comment]
Some notes about the gesture detectors, since I was just digging into them:

  * TextSelectionGestureDetector
    - Detects gestures relevant to text editing at a low level.
  * TextSelectionGestureDetectorBuilder
    - The main purpose of this class is to reuse some text editing business logic across the different text editing widgets (TextField, CupertinoTextField, TextFormField, SelectableText).
    - It creates an instance of TextSelectionGestureDetector with buildGestureDetector.
    - It has some built-in handlers for these gestures, but allows the text editing widgets to override them.
    - This class is probably obsolete now that we have Shortcuts/Intents/Actions, and I think I plan to get rid of it in the long term.

Now about the single/double taps, I think it may have originally been written this way to match Android's behavior, but based on some quick testing, it's incorrect for most other platforms.  When tapping repeatedly:

On Android Keep, the selection alternates between the word and collapsed at the tap location.
On iOS Notes, the word is selected and then the selection doesn't change.
On Mac TextEdit, the word is selected, then the paragraph is selected, and then selection doesn't change.

I'm going to try to convert onSingleTapUp to Shortcuts/Actions/Intents as the first example of a mouse/tap gesture.  I'll reference the PR here if I do.

**Comment by venkatd** [comment]
Ok great!

> This class is probably obsolete now that we have Shortcuts/Intents/Actions, and I think I plan to get rid of it in the long term.
Makes sense. I suppose you can have different mappings of Gesture/Intent/Action depending on the platform. Like how the text editing now works.

I can first focus on adding triple click to the gesture detector, then can copy what you come up with to link it up to an intent/action to select the paragraph.

**Comment by matthew-carroll** [comment]
@justinmc 
 - Here is an implementation of a `GestureDetector` that recognizes single, double, and triple taps: https://github.com/superlistapp/rich_text_editor/blob/main/lib/src/infrastructure/multi_tap_gesture.dart

Also:
 - Is there a reason that you want to bake actions/intents into low level text operations? 
 - Is it a good idea to add more selection calculations to `RenderParagraph`?

As we've discussed before, the diversity of abilities that are combined in Flutter's text implementation is prohibitive. I'm concerned about more being added. `RenderParagraph` is the last bastion of a reasonably composable piece within the entire text implementation. If that gets bloated with selection calculations and/or a higher level action/intent/response structure, then there is no place for developers to hide from the monolith. As you know, I've already had to re-implement everything above `Text` for an editor. I'd really like to see the Flutter org prioritize some concern for the lack of composability in this area.

WRT calculating a paragraph selection based on gestures, where does that categorization end? This thread talks a lot about standard platform behaviors, but Flutter's ability to select text cannot be limited to what platforms do by default. Developers need control over this. If I want to select a single sentence on triple-click instead of a full paragraph, I need to be able to do that. Moreover, what constitutes a paragraph? Is it all the content in the `Text` widget, or is it the content between 2 newlines, or something else? The platform standards should be *achievable* but they should not be the basis for new APIs, IMO.

I would encourage some deep thought before adding any more stuff to the lowest level framework-side APIs. I would also encourage you and any other team members to come take a look at this text editor project to get a feel for how different the world of desktop use-cases might be compared to mobile.

I think I've proven that pretty much any desired text ability is achievable from outside the framework, if developers put in the work. The only existential risk for text-based customization is the possibility that Flutter injects overly-constrained APIs into `Text` and `RenderParagraph`. I would very much encourage a "do no harm" approach in that regard over a "provide all the expected defaults" approach that risks the former. If developers need more control over what already exists then I think the obvious answer is to break things apart, not pile more on top.

As always, you're welcome to reach out if you'd like to discuss thoughts on improving this area of Flutter.

**Comment by lukepighetti** [comment]
I'll just take a moment to say that I agree that Flutter needs to have less opinionated base APIs. But it also needs a middle layer that makes them perform 'as expected.' Right now it tries to do both and it's not working well. And I honestly don't believe the community has what it takes to maintain the fidelity on the middle layer in the long run. Even if someone does solve the 'middle layer' for individual pieces, like text manipulation, the bus factor will be very high, and it won't necessarily jive well with other 'middle layer' solutions.

**Comment by justinmc** [comment]
I think @matthew-carroll, @lukepighetti, and I are actually all more or less in agreement.

  * RenderEditable should be a strict and generic text editing model.
  * Business logic (like what a Mac does on triple click) should live elsewhere (@lukepighetti's middle layer, my [DefaultTextEditingActions](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/default_text_editing_actions.dart) and [DefaultTextEditingShortcuts](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/default_text_editing_shortcuts.dart)).

So something like triple click would work like this:

  * A triple click is detected and a generic Intent is dispatched (e.g. TripleClickTextIntent).  This way, whatever widget is doing the gesture detection doesn't have to know anything about what the gesture does.
  * TripleClickTextIntent gets mapped to the correct business logic based on the platform in [DefaultTextEditingActions](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/default_text_editing_actions.dart).  This is like @lukepighetti's middle layer, if I understand correctly.
  * RenderEditable is called to perform the desired behavior through generic text editing methods like selectParagraph or selectSentence.

With this pattern, I am trying to make RenderEditable even simpler than it is now.  It shouldn't have to do any gesture detection itself anymore (I'm in the process of getting rid of its [RawKeyboard listener](https://github.com/flutter/flutter/blob/23015dc501d8ab206036bc9ff761f24af8c270b3/packages/flutter/lib/src/rendering/editable.dart#L2930)).  It also shouldn't need to know anything about specific platforms or how they behave (I hope to remove all uses of defaultTargetPlatform from it).

Also, all of this text editing behavior applies globally, not tied to widgets like TextField like it is now, and it's easily overridable by the Flutter developer by using an Actions widget.

Here's my design doc on this idea for those that haven't seen it: https://docs.google.com/document/d/1QaVIr1bbOWJGNyyNsB75sXgTDRMuxP268sLGTBNDut0/edit?usp=sharing

**Comment by venkatd** [comment]
I may be splitting hairs here, but would it make sense to call the intent `SelectParagraphIntent` or `SelectWordIntent`? Triple click feels more a shortcut in the Shortcuts/Intents/Actions triad.

So in the more general form it's like: event (keyboard shortcut, mouse gesture, voice command, shaking the phone, etc) => intent => action

When I double click somewhere, my intent may be to select that word. If I triple click, my intent may be to select the line. Or it may be to select the paragraph.

**Comment by justinmc** [comment]
I agree that this naming is a bit weird compared with how the shortcuts are done, but I think it's really important to keep the widget that invokes it ignorant about what it will do.

The PR that I promised to link that sets this up is [here](https://github.com/flutter/flutter/pull/79414) (WIP), and the part where I map the Intent to different Actions is [here](https://github.com/flutter/flutter/pull/79414/files#diff-a7ab0a5db9b1cb96f1d25b62073b4d8bbbe0f4dfa56578bede557c528cd0c121R41).

**Comment by sgehrman** [comment]
Critical feature.  Why hasn't this been fixed yet?

**Comment by justinmc** [comment]
We are building clean a way to do this and other mouse interactions via Actions (CC @LongCatIsLooong) which unfortunately is going to take some time.  We could hack it in the meantime, but I'd like to avoid that if possible.

**Comment by ryzizub** [comment]
Flutter should behave just like native platforms. I would expect such a bug to have higher priority as I classify it as a critical issue

**Comment by justinmc** [comment]
Mentioned in https://github.com/flutter/flutter/issues/97866#issuecomment-1031448768, we should also support quadruple click and triple/quadruple click + drag.

**Comment by sgehrman** [comment]
This is the one and only feature that makes Flutter Web unusable.  No one is going to use a website with non standard funky text fields.  Please fix ASAP.

**Comment by codeXLinkX** [comment]
I wonder what the community can do to bump this issue up in the backlogs list.

**Comment by Renzo-Olivares** [comment]
I’ll be taking this on as one of my priorities. It depends on a refactor but I am working on it. I’ll post updates here as I progress.

**Comment by umitk34** [comment]
> I’ll be taking this on as one of my priorities. It depends on a refactor but I am working on it. I’ll post updates here as I progress.

Hey, @Renzo-Olivares I would like to hear about your progress so far?

**Comment by Renzo-Olivares** [comment]
Hey @umitk34 I'm currently working on a proof of concept and design doc and will post it here as soon as it's ready. Thank you for the interest.

**Comment by djsjr** [comment]
@Renzo-Olivares Any update on this?

**Comment by Renzo-Olivares** [comment]
Hey @djsjr, thanks for your interest in this project. I have a proof of concept PR as well as a design doc [here](https://github.com/flutter/flutter/pull/105165). I'm currently working on some related work https://github.com/flutter/flutter/pull/109573 which is building a `GestureRecognizer` that can handle taps and drags for text selection (e.g. double tap to drag, triple tap to drag). Once that is done, I can continue moving this project forward.

**Comment by ngugcx** [comment]
> Hey @djsjr, thanks for your interest in this project. I have a proof of concept PR as well as a design doc [here](https://github.com/flutter/flutter/pull/105165). I'm currently working on some related work #109573 which is building a `GestureRecognizer` that can handle taps and drags for text selection (e.g. double tap to drag, triple tap to drag). Once that is done, I will continue moving this project forward.

If double tap/triple tap enabled, there is a delay issue (#110300) which make user experience awful.
Does Flutter have the plan to polish these basic controls according to the desktop usage habits?

**Comment by Renzo-Olivares** [comment]
@ngugcx I am working on a new `TapAndDragGestureRecognizer`, where that delay is not an issue.

**Comment by seel-channel** [comment]
A solution to this issue

```dart
bool _canSelectAll = false;
final TextEditingController _controller = TextEditingController();
Timer? _selectAllTimer;

void _onTap() {
  _selectAllTimer?.cancel();
  _selectAllTimer = null;

  if (_controller.selection.isCollapsed) {
    _canSelectAll = true;
    _selectAllTimer = Timer(const Duration(milliseconds: 500), () {
      _canSelectAll = false;
    });
    return;
  }

  if (!_canSelectAll) return;

  final selection = TextSelection(
    baseOffset: 0,
    extentOffset: _controller.document.length,
  );

  _controller.selection = selection;
  _canSelectAll = false;

  return;
}



@override
Widget build(BuildContext context) {
  return TextField(
    onTap: _onTap,
  );
}
```

**Comment by bionascu** [comment]
Hi @Renzo-Olivares! Any chance there has been some progress on this? Many thanks!! 🙏

**Comment by Renzo-Olivares** [comment]
Sure thing @bionascu!

In terms of just closing this issue "Supporting select all text on triple click" I am currently working on introducing a `ParagraphBoundary` subclass in #116549 that will allow for this behavior. That PR is centered around keyboard actions but the boundary subclass will be reusable. Once that is merged I'll start another PR to add triple tap to select for tap/clicks and that should close this issue.

Updates on other touch/mouse text gestures:

As of recent #117502 was merged that added support for double tap/click + drag to select text word by word. It also added support for dragging to move the cursor on mobile platforms.

#101151 is the larger project that aims to make these tap/click gestures overridable. This is still in progress and was waiting on #117502 to get merged.

**Comment by reza00farjam** [comment]
@Renzo-Olivares , Hi!
I faced this issue even in the latest release, did it fixed somewhere else?

**Comment by Renzo-Olivares** [comment]
@reza00farjam it is fixed in master.
